### ISSUE TITLE
Errors in reverse sparsity prop in Integrator

### ISSUE THREAD

**Issue by jgillis** [casadi/casadi#1577]
Below is an excerpt from the updated demo notebook.
It features a nasty regression bug: see the results of derivative_test.
```python
from casadi import *
from casadi.tools import *

t  = SX.sym("t")        # time
u  = SX.sym("u")        # control
x  = SX.sym("[p,q,c]")  # state
p,q,c = x

ode = vertcat([(1 - q**2)*p - q + u, p, p**2+q**2+u**2])
print ode, ode.shape

f = SXFunction("f",daeIn(x=x,t=t,p=u),daeOut(ode=ode))

f.printDimensions()

tf = 10.0
N = 20
dt = tf/N

Phi = Integrator("Phi","cvodes",f,{"tf":dt})

x0 = vertcat([0,1,0])

print Phi({"x0":x0})

X = struct_symMX([
     (
      entry("x", repeat=N+1, struct=struct(["p","q","c"]) ),
      entry("u", repeat=N)
     )
])

Xf = Phi( x0=X["x",0],p=X["u",0] )["xf"]

print Xf

g = [] # List of constraint expressions

for k in range(N):
  Xf = Phi( x0=X["x",k],p=X["u",k] )["xf"]
  g.append( X["x",k+1]-Xf )

obj = X["x",N,"c"] # c_N

nlp = MXFunction("nlp",nlpIn(x=X), nlpOut(g=vertcat(g),f=obj) )

print nlp

jacG = nlp.jacobian("x","g")

S = jacG.outputSparsity(0)

print S.shape

DMatrix.ones(S)[:20,:20].sparsity().spy()

solver = NlpSolver("solver","ipopt",nlp,{"derivative_test":"first-order"})

lbx = X(-inf)
ubx = X(inf)

lbx["u",:] = -1; ubx["u",:] = 1   #   -1 <= u(t) <= 1

lbx["x",0] = ubx["x",0] = x0      # Initial condition

sol_out = solver({
           "lbg": 0, # Equality constraints for shooting constraints
           "ubg": 0, #    0 <= g <= 0
           "lbx": lbx,
           "ubx": ubx
})
```

**Comment by jgillis** [comment]
Also, expanding this triggers a segfault as per #1552.

**Comment by jgillis** [comment]
Cornering error: only appears with "ad_weight_sp": 1 (reverse mode)

**Comment by jgillis** [comment]
reverse mode sparsity seems to find a few false-zeros.

**Comment by jgillis** [comment]
The constraint Jacobian of the reverse mode has this structure:
```
**.**..............................................................................
**.*.*.............................................................................
**.*..*............................................................................
....**.**..........................................................................
....**.*.*.........................................................................
....**.*..*........................................................................
```

For forward mode it is:
```
**.**..............................................................................
**.*.*.............................................................................
****..*............................................................................
....**.**..........................................................................
....**.*.*.........................................................................
....****..*........................................................................
```

note the extra nonzeros is $dc_next/dc$.
Numerically, the matrices differ in the $dc_next/dp$ and  $dc_next/dnext$ place.

**Comment by jgillis** [comment]
Reduced to:
```
from casadi import *
from casadi.tools import *

t  = SX.sym("t")        # time
u  = SX.sym("u")        # control
x  = SX.sym("[p,q,c]")  # state
p,q,c = x

ode = vertcat([(1 - q**2)*p - q + u, p, p**2+q**2+u**2])
print ode, ode.shape

f = SXFunction("f",daeIn(x=x,t=t,p=u),daeOut(ode=ode))

f.printDimensions()

tf = 10.0
N = 20
dt = tf/N

for i in [0,1]:
	print "mode",i
	Phi = Integrator("Phi","rk",f,{"tf":dt,"ad_weight_sp":i})
	Phi.jacSparsity("x0","xf").spy()
```

```
mode 0
**.
**.
***
mode 1
**.
**.
**.
```

**Comment by jaeandersson** [comment]
Unrelated:
```python
x  = SX.sym("[p,q,c]")  # state
p,q,c = x
```
should be replaced with:
```python
p = SX.sym("p")
q = SX.sym("q")
c = SX.sym("c")
x = vertcat((p,q,c))  # state
```

**Comment by jaeandersson** [comment]
More isolated:
```python
from casadi import *

p = SX.sym("p")
q = SX.sym("q")
c = SX.sym("c")
f = SXFunction("f",daeIn(x=vertcat((p,q,c))),daeOut(ode=vertcat((q, p, p))))

for i in [0,1]:
    print "mode",i
    Phi = Integrator("Phi","rk",f,{"ad_weight_sp":i})
    Phi.jacSparsity("x0","xf").spy()
```

**Comment by jaeandersson** [comment]
Most likely work vector related. Will start a git bisect to isolate.

**Comment by jaeandersson** [comment]
Heh, not very surprising the first bad commit is 8db407be698776c7798cdef56f0eb73dca1df4d0.

**Comment by jaeandersson** [comment]
Fixed in edb4038a0b4a95af597d3610193d84893bad48eb.

**Comment by jgillis** [comment]
Todo(joris): unittest
